### ISSUE TITLE
Remote compile and run C++ code

### ISSUE THREAD

**Issue by ghost** [microsoft/vscode-cmake-tools#603]
Is it possible to configure cmake tools to compile, build and run on remote device when ssh, sftp available?

**Comment by vector-of-bool** [comment]
Not currently. This is a feature I've wanted to look into, though, especially as I do a lot of development via Docker.

I'd be interested in how this would work with cpptools, though. I do not know if they respect remote filesystems at the moment.

**Comment by wcwagner** [comment]
@vector-of-bool I develop in docker currently as well. I can use vscode-cpptools but I have to create a docker volume to map my files back to the host. However, is there any way to prefix build commands with `docker exec -it ...`? I believe I saw you mention that this would be incompatible with cmake-server somewhere though, but I can't find where that was.

**Comment by TheErk** [comment]
When using docker there is two options:

1) you use docker as a "cross-target" and you cross-compile from the host, thus you speak to the CMake server on the host, and cmake is executed on the host, compilation as well etc...

2) you use docker as a "native-target" so beside executing vscode on the host all other dev process: running cmake, cmake server, compiling etc... is done in the docker.

I personally use 2) and since there is no "in-docker" support with CMakeTools I do run vscode in the docker as well :-)
I don't think 1) is useful when using docker.

This works perfectly well. In this scheme I install an ssh server in the docker container, then I ssh into it and let ssh ForwardX11 the display to the host I'm ssh-ing from.
This is a easy setup because vscode+CMakeTools can talk to cmake-server (https://cmake.org/cmake/help/v3.7/manual/cmake-server.7.html) "as usual".

If you want to keep vscode outside docker you should find a way to make CMakeTools talk to cmake-server named pipe in the docker.

**Comment by TheErk** [comment]
This should be [theoretically] possible if the named pipe is in a docker mounted volume which thus
can be seen outside (for vscode/cmaketools) and inside the docker (for cmake-server).

**Comment by bobbrow** [comment]
I suspect it would take a bit of work to implement the vscode.FileSystemProvider API.  We started tagging remote filesystem issues to try to get an idea as to how urgent the need is for first-class support.  We only support `file:///` URIs right now, so we currently require users to mount/map their remote filesystems before using our extension.

I have heard mention of [this extension](https://marketplace.visualstudio.com/items?itemName=Kelvin.vscode-sshfs), but haven't played around with it yet.

**Comment by TheErk** [comment]
Even if I use ssh go into my docker container. May be requiring ssh is too much?
I tend to think that when your do development your source are usually mapped from host to docker through a mounted docker volume already.  The only weird thing is you usually "see" the  very same source at different absolute path inside docker as you see them outside docker.

May be that if cpptools was simply "aware" of the remapping of the source you may work with the source on the host directly? This is a wild guess.

There is a bunch of "remote editing" extension for vscode, see e.g.: [remote fs](https://marketplace.visualstudio.com/items?itemName=liximomo.remotefs) for example.
I did not tested [yet].

I'm no vscode extension developer so I don't know whether if this uses vscode.FileSystemProvider API, but may be cpptools could use one of those extension.

**Comment by Yanpas** [comment]
I use docker too. To workaround running cmake inside container I've created dk-cmake command. But there is no way to specify dk-ninja command

**Comment by JerryJoyce** [comment]
I'd be very glad to see this implemented.  I do all my builds inside a docker container.

**Comment by DoDoENT** [comment]
I would also like this feature. I usually develop on MacOS, but sometimes some I need to debug some Linux-specific part of my code. At the moment I SSH into my Linux server and do minimal code changes in vim while compiling and running with cmake manually. And debugging via command-line GDB via SSH is a nightmare.

It would be great if VSCode had the ability similar to [this one in full Visual Studio](https://devblogs.microsoft.com/cppblog/using-visual-studio-for-cross-platform-c-development-targeting-windows-and-linux/), which enables full-scale development of your code on Windows while compiling and debugging on remote Linux machine.

**Comment by mika-fischer** [comment]
This is now available in vscode: https://code.visualstudio.com/blogs/2019/05/02/remote-development

However, CMake Tools hangs with `CMake Tools initializing...: Creating backend` for me...

**Comment by PhilipMourdjis** [comment]
I get the same issue on Windows, works fine on Ubuntu 18.04

**Comment by exbutterfly** [comment]
Extension is blocked on calling "Rollback opt-in" which not appear on Remote-SSH (i don't understand why)

Try `"cmake.rollbar-optin3": false` in `settings.json` config file. It help me.

[Here](https://github.com/vector-of-bool/vscode-cmake-tools/issues/704#issuecomment-508588877) is a similar issue.

**Comment by bytesoftly** [comment]
Any updates on this in 2020, I'd love to be able to use cmake-tools with VSCode remote sessions?

**Comment by andreeis** [comment]
It should work with the latest CMake. Reactivate if you encounter any other issues.

**Comment by definelicht** [comment]
I tried using CMake debugging in a remote session, but I'm getting this error message:

![target_name_undefined](https://user-images.githubusercontent.com/2845320/106164021-d8c2f000-6189-11eb-9184-0e83a25816a4.gif)

As you can see in the toolbar, the target is set to `dispatch`. Any idea what could be causing this?

**Comment by bobbrow** [comment]
[dispatch] is the build target.  It's unclear why the extension didn't ask you to select a debug target.  Can you open a new issue with more details about your project?

**Comment by definelicht** [comment]
Indeed, I just needed to manually configure the CMake Build Target. Thank you, it now correctly launches the specified debug target.
